# ITS_Database_Outgroups
Comparing "standard" UNITE taxonomic assignments with UNITE+Outgroups in published fungal ITS1 studies

_____

## **Workflow:**

```
bash ./01_download_data_and_make_ps.sh
bash ./02_combine_and_analyze.sh
```

#### The first BASH script runs the following 3 R scripts to generate a separate phyloseq object for each taxonomic assignment method and each study listed in "./metadata/Study_List.xlsx"

1. Download sequence data from all SRA accessions listed in "./metadata/Study_List.xlsx" keeping only FWD reads
    + "./R/Download_Fastqs.R"
2. Extract ITS1 region
    + "./R/Extract_ITS1.R"
3. Process each project with DADA2, assign taxonomy with each database, save phyloseq objects
    + "./R/make_phyloseq_objects.R"

#### The second BASH script runs the following R scripts to combine and analyze the taxonomic assignments generated by the first BASH script

1. Tally the number and proportion of kingdom-level taxonomic assignments from each study and builds separate data frames for each
    + ".R/parse_kingdom_assignments.R"
2. Combine all study data frames, build plots, model, and test
    + "./R/combine_and_analyze_assignments.R"

_____

All analyses conducted with:
+ R verion 3.4.4
+ R-studio version 1.2.1
+ GNOME Terminal 3.18.3 using VTE version 0.42.5 +GNUTLS
