# ITS_Database_Outgroups
The purpose of this project is to compare "standard" UNITE taxonomic assignments with UNITE+Outgroups in published fungal ITS1 studies.



_____

## **Workflow:**

```
bash ./01_download_data_and_make_ps.sh
bash ./02_combine_and_analyze.sh
```

#### The first BASH script runs the following 3 R scripts to generate a separate phyloseq object for each taxonomic assignment method and each study listed in "./metadata/Study_List.xlsx"

1. Download sequence data from all SRA accessions listed in "./metadata/Study_List.xlsx" keeping only FWD reads
    + "./R/Download_Fastqs.R"
2. Extract ITS1 region
    + "./R/Extract_ITS1.R"
3. Process each project with DADA2, assign taxonomy with each database, save phyloseq objects
    + "./R/make_phyloseq_objects.R"  #cange script to remove ps objects once written to RDS file
    
#### The second BASH script runs the following R scripts to combine and analyze the taxonomic assignments generated by the first BASH script

1. Tally the number and proportion of kingdom-level taxonomic assignments from each study and builds separate data frames for each
    + ".R/parse_kingdom_assignments.R"
2. Combine all study data frames
    + "./R/combine_taxonomic_data.R" #rename script to match
3. Calculate alpha and beta diversity of fungi from each tax_table (using pyloseq RDS objects as input
    + Read in all 3 ps RDS files from a given study
    + Remove non-fungi
    + Get alpha and beta diversity estimates (specnumber, Shannon, Simpson, Chao1, betadisper)

```
# alpha diversity is simple enough...

# mean beta-dispersion
dis <- vegdist(otu_table(ps))
groups <- sample_data(ps)$BioProject
mod <- betadisper(dis, groups)
mean(mod$distances)

# comparing beta-diversity between taxonomy treatments
library(tidyverse)

# total species counts for all samples
v1 = colSums(otu_table(ps_UNITE))
v2 = colSums(otu_table(ps_ALL))
v3 = colSums(otu_table(ps_NCBI))

# make data frame (one row for each method, cols are species, values are sum of observations)
d1 = data.frame(Total = v1,Spp = names(v1),Method = "UNITE")
d2 = data.frame(Total = v2,Spp = names(v2),Method = "UNITE_ALL")
d3 = data.frame(Total = v3,Spp = names(v3),Method = "UNITE_NCBI")
df = rbind(d1,d2,d3)
df = spread(df,key=Spp,value = Total)
row.names(df) <- df$Method
df <- df %>% select(-Method)
df[is.na(df)] <- 0

# have list of distance matrices tat grows for each study
df.dist = vegdist(df)
dist.list <- list(StudyID=df.dist)

# Same for permanova
Database = row.names(df)
list(StudyID = adonis(df ~ Database))
```

4. Add point estimates to dataframe (grows with each iteration of loop)
    + Data Frame (StudyID, Database, Shannon, Richness, meanbetadispersion)
    + Data Frame (StudyID, list(dist_matrix), list(adonis_table))
    + Remove ps objects and move to next study
    + After loop close, save data frames as RDS
5. Plotting
    + Read in data frame of diversity estimates from previous script
    + Plot alpha diversity (difference between databases ~ distance from Estonia?) ... by ecosystem?
    + Plot Beta Diversity (same as above)
    + Extract P-values and R2 from adonis
    + Compare distance matrices

```
l = list(Study1 = df.dist, # the list of distance matrices from previous script's dataframe
         Study2 = df.dist2)
d = data.frame(Method = c("UNITE","UNITE_ALL","UNITE_NCBI"), lapply(l,c)) # add Ecosystem, LatLon, etc
long = gather(d,key = Study,value = Distance,2:dim(d)[2])


```

6. Modeling
    + ANOVA


_____

All analyses conducted with:
+ R verion 3.4.4
+ R-studio version 1.2.1
+ GNOME Terminal 3.18.3 using VTE version 0.42.5 +GNUTLS
